File: py3dpolys_le/plot_hic.py
#! /usr/bin/env python

import argparse
import logging
import os
import sys

import h5py
import numpy as np
from matplotlib import pyplot as plt
from matplotlib.ticker import FuncFormatter
import glob
import cooler
import dask.array as da
from scipy.signal import savgol_filter, find_peaks, find_peaks_cwt

# Initialization
logger = logging.getLogger(__name__)

DEFAULT_CMAP = "hot_r"
# tuned for simulation HiCs
DEFAULT_RESOLUTION = 2000
DEFAULT_HIC_WILDCARD = "hic*.hdf5"
DEFAULT_CLIM = [-2.75, 0]  # found to be mostly OK for our experimental and simulation HiCs
DEFAULT_PLOT_FORMAT = "png"
CHR_X_SYNONYMS = ['6', 'chrX', 'X']
DEFAULT_TITLE = "Hi-C for measurement {hic_file}"
DEFAULT_DPI = 150

# default simulation resolution
SIM_RESOLUTION = 2000

# experimental: extra optional
crop = None  # [9000//2, 15000//2]


def cli_parser():
    p = argparse.ArgumentParser()
    p.add_argument("-o", "--output_folder", default=".",
                   help="'Analysis' step output folder containing raw hic_*.hdf5 files.")
    p.add_argument("-w", "--hic_wildcard", default=DEFAULT_HIC_WILDCARD,
                   help="Hi-C files wildcard. Supported hic formats: *.hdf5 (simulation's format), *.cool, *.mcool")
    p.add_argument("--hic_chrs", nargs='*', default=CHR_X_SYNONYMS,
                   help="Synonyms of the chromosome from Hi-C matrixes to be ploted. "
                        "Default: synonyms for chrX in c.elegans: chrX, X, 6 ")
    p.add_argument("-r", "--resolution", default=DEFAULT_RESOLUTION, type=int, help="Hi-C data resolution in bp.")
    p.add_argument("-b", "--balanced", action='store_true', help="Get balanced if available, for .cool and .mcool.")
    p.add_argument("-c", "--cmap", default=DEFAULT_CMAP,
                   help="Color map: cool, hot_r, gist_heat_r, afmhot_r, YlOrRd, Greys, gist_yarg. "
                        f"Default: {DEFAULT_CMAP}")
    p.add_argument("--clim", nargs='*', type=float, default=DEFAULT_CLIM,
                   help="Set the color limits of the current image, see matplotlib.pyplot.clim parameter."
                        f"Default: {DEFAULT_CLIM[0]}, {DEFAULT_CLIM[1]}")
    p.add_argument("-f", "--plot_format", default=DEFAULT_PLOT_FORMAT,
                   help=f"Image file format extension: png, tif, svg. Default: {DEFAULT_PLOT_FORMAT}")
    p.add_argument("--title", default=DEFAULT_TITLE,
                   help="Plot title. It could be template containing parameters in {}: hic_file, output_folder, "
                        "resolution, balanced"
                        f". Default: {DEFAULT_TITLE}")
    return p


def get_hic(hic_file, resolution, balanced, hic_chrs):
    if hic_file.endswith('.hdf5'):
        # directly from .hdf5
        with h5py.File(hic_file, 'r') as f:
            a_group_key = list(f.keys())[0]
            logger.info(f"get_hic for {hic_file} HDF5 Keys {f.keys()} and dimentions {a_group_key}")
            # Get the data
            data = list(f[a_group_key])
            hic = np.array(data)
            np.fill_diagonal(hic, 0)
            if resolution > SIM_RESOLUTION:  # need binning
                hic = hic_coarsen(hic, hic_file, resolution)
            res = SIM_RESOLUTION
    elif hic_file.endswith('.cool') or hic_file.endswith('.mcool'):
        # use the root cool
        cooler_ref = f'{hic_file}::/' if hic_file.endswith('.cool') else f'{hic_file}::/resolutions/{resolution}'
        hic_cooler = cooler.Cooler(cooler_ref)
        hic_chr = list(set(hic_cooler.chromnames) & set(hic_chrs))[0]
        balance = (hic_cooler.bins()['weights'] is not None) if balanced else balanced
        hic = hic_cooler.matrix(balance=balance).fetch(hic_chr)
        if balanced:
            hic = np.nan_to_num(hic)  # nan -> 0
        np.fill_diagonal(hic, 0)
    return hic


def hic_coarsen(hic, hic_h5, resolution, from_resolution=SIM_RESOLUTION):
    factor = resolution // from_resolution
    if factor == 1:
        return hic   # no coarsen
    if resolution % from_resolution != 0:
        logger.error(f'Cannot do binning from resolution {from_resolution} to {resolution} '
                     f'for hic file {hic_h5}')
    # coarsen
    dhic = da.from_array(hic)  # , chunks=hic.shape[0] // factor)
    dhic_binned = da.coarsen(np.sum, dhic, {0: factor, 1: factor})
    hic = dhic_binned.compute()
    return hic


def run(output_folder, hic_wildcard=DEFAULT_HIC_WILDCARD, resolution=DEFAULT_RESOLUTION, balanced=False,
        hic_chrs=CHR_X_SYNONYMS, cmap=DEFAULT_CMAP, clim=DEFAULT_CLIM, title=DEFAULT_TITLE,
        plot_format=DEFAULT_PLOT_FORMAT):
    logger.info(f'Plotting HiC for {output_folder} with color map: {cmap} in file format: {plot_format} ...')
    if not cmap:
        cmap = DEFAULT_CMAP
    if not plot_format:
        plot_format = DEFAULT_PLOT_FORMAT

    hic_files = sorted(glob.glob(os.path.join(output_folder, hic_wildcard)))
    for hic_file in hic_files:
        hic = get_hic(hic_file, resolution, balanced, hic_chrs)
        print(f"data.shape: ${hic.shape}")

        fig = plt.figure()

        # crop desired section
        if crop:
            if len(crop) == 2:
                hic = hic[crop[0]:crop[1], crop[0]:crop[1]]
            else:
                hic = hic[crop[0]:, crop[0]:]

        hic_log = np.log10(hic)

        res_kb = resolution/1000
        plt.gca().get_xaxis().set_major_formatter(FuncFormatter(lambda x, p: int(x*res_kb)))
        plt.gca().get_yaxis().set_major_formatter(FuncFormatter(lambda y, p: int(y*res_kb)))

        plt.imshow(hic_log, interpolation='nearest', cmap=cmap)  # color scale
        # cbar_h = plt.colorbar()

        # # START: loop anchors finder: for experimental HiCs
        # hic_aggregated = np.log10(hic.sum(axis=0)) * -1000
        # hic_aggregated[hic_aggregated == -np.inf] = 0
        # hic_aggregated[hic_aggregated == np.inf] = 0
        # hic_aggregated_smooth = savgol_filter(hic_aggregated, 7, 3) + 700
        # plt.plot(hic_aggregated_smooth)
        # # peaks, _ = find_peaks(1000-hic_aggregated_smooth, height=170, width=6)  # np.arange(10, 20))
        # # resolution: 20000
        # #peaks = find_peaks_cwt(-hic_aggregated_smooth, widths=np.arange(5, 30))
        # peaks = find_peaks_cwt(-hic_aggregated_smooth, widths=np.arange(5, 27))
        # plt.plot(peaks, hic_aggregated_smooth[peaks], "x")
        # # END: loop anchors finder

        print(plt.rcParams['axes.prop_cycle'].by_key()['color'])

        cbar_h = plt.colorbar()
        # mu = np.mean(hic)
        # sd = np.std(hic)
        # qt = np.quantile(hic, 0.50)
        # plt.clim(np.log10(mu - 1.64*sd), np.log10(mu + 1.64*sd))
        # plt.clim(np.log2(mu + 1.25*sd), 0)
        if clim and len(clim) > 1:
            plt.clim(clim[0], clim[1])
        hic_file_clean = hic_file.replace('_hic_003.hdf5', '').replace('./', '')   # extra clean
        plt.title(title.format(hic_file=hic_file_clean, output_folder=output_folder, resolution=resolution, balanced=balanced))

        # cbar_h.ax.tick_params(labelsize=11)
        # plt.show()

        basename = os.path.basename(hic_file)
        extension = basename.split(".")[-1]
        crop_ext = f'_c{crop[0]}{f"-{crop[1]}" if len(crop)==2 else ""}' if crop else ""
        plot_file = os.path.join(output_folder, basename.replace(f'.{extension}', f'{crop_ext}_{cmap}.{plot_format}'))

        fig.savefig(plot_file, dpi=DEFAULT_DPI)
        plt.close()
        print(f'Hic plot saved in file {plot_file}')


def main():
    args = cli_parser().parse_args(sys.argv[1:])

    run(args.output_folder, hic_wildcard=args.hic_wildcard, resolution=args.resolution, balanced=args.balanced,
        hic_chrs=args.hic_chrs, cmap=args.cmap, clim=args.clim, title=args.title, plot_format=args.plot_format)


if __name__ == '__main__':
    main()

File: src/PlotHic.jl
"""
Plot Hi-C module
"""
module PlotHic

import Base.CoreLogging: @info, @warn, @error, @debug
using HDF5
using Plots
using Glob
using PyCall
using ArgParse
using LinearAlgebra
using Base: log10

# Use Julia cooler implementation
include(joinpath(@__DIR__, "cooler", "CoolerModule.jl"))
using .CoolerModule
@info "Using Julia cooler implementation"

const DEFAULT_CMAP = "hot_r"
const DEFAULT_RESOLUTION = 2000
const DEFAULT_HIC_WILDCARD = "hic*.hdf5"
const DEFAULT_CLIM = [-2.75, 0]
const DEFAULT_PLOT_FORMAT = "png"
const CHR_X_SYNONYMS = ["6", "chrX", "X"]
const DEFAULT_TITLE = "Hi-C for measurement {hic_file}"
const DEFAULT_DPI = 150
const SIM_RESOLUTION = 2000

# Helper function to fill diagonal with a value (equivalent to np.fill_diagonal)
function fill_diagonal!(matrix::Matrix, value::Number)
    n = min(size(matrix, 1), size(matrix, 2))
    for i in 1:n
        matrix[i, i] = value
    end
end

# Map matplotlib colormap names to Plots.jl colormap symbols
# Plots.jl uses different names and doesn't support "_r" suffix for reversed colormaps
function get_plots_colormap(cmap::String)::Symbol
    # Map common matplotlib colormaps to Plots.jl equivalents
    colormap_map = Dict(
        "hot_r" => :heat,           # reversed hot -> heat (similar appearance)
        "hot" => :heat,
        "gist_heat_r" => :heat,
        "afmhot_r" => :heat,
        "YlOrRd" => :YlOrRd,        # Available in Plots.jl
        "Greys" => :grays,
        "gist_yarg" => :grays,
        "cool" => :cool,
        "viridis" => :viridis,
        "plasma" => :plasma,
        "inferno" => :inferno,
        "magma" => :magma,
        "YlGnBu_r" => :YlGnBu,      # reversed -> normal
        "YlGnBu" => :YlGnBu,
    )
    
    # Check if we have a direct mapping
    if haskey(colormap_map, cmap)
        return colormap_map[cmap]
    end
    
    # Try to convert to Symbol directly (might work for some colormaps)
    try
        return Symbol(cmap)
    catch
        # Fallback to heat if unknown
        @warn "Unknown colormap '$cmap', using 'heat' as fallback"
        return :heat
    end
end

function get_hic(hic_file::String, resolution::Int, balanced::Bool, hic_chrs::Vector{String})::Matrix
    if endswith(hic_file, ".hdf5")
        h5open(hic_file, "r") do f
            a_group_key = first(keys(f))
            @info "get_hic for $hic_file HDF5 Keys $(keys(f)) and dimensions $a_group_key"
            data = read(f[a_group_key])
            hic = Array(data)
            fill_diagonal!(hic, 0)
            if resolution > SIM_RESOLUTION
                hic = hic_coarsen(hic, hic_file, resolution)
            end
            return hic
        end
    elseif endswith(hic_file, ".cool") || endswith(hic_file, ".mcool")
        cooler_ref = endswith(hic_file, ".cool") ? "$hic_file::/" : "$hic_file::/resolutions/$resolution"
        hic_cooler = CoolerFile(cooler_ref)
        hic_chr_names = chromnames(hic_cooler)
        hic_chr = first(intersect(hic_chr_names, hic_chrs))
        balance = balanced && (bins(hic_cooler)["weights"] !== nothing)
        mat_obj = matrix(hic_cooler, balance=balance)
        hic = fetch(mat_obj, hic_chr)
        if balanced
            hic = replace(hic, NaN => 0.0)
        end
        fill_diagonal!(hic, 0)
        return hic
    else
        error("Unsupported file format: $hic_file")
    end
end

function hic_coarsen(hic::Matrix, hic_h5::String, resolution::Int, from_resolution::Int = SIM_RESOLUTION)::Matrix
    factor = resolution ÷ from_resolution
    if factor == 1
        return hic
    end
    if resolution % from_resolution != 0
        @error "Cannot do binning from resolution $from_resolution to $resolution for hic file $hic_h5"
    end
    
    N = size(hic, 1)
    new_N = N ÷ factor
    hic_binned = zeros(new_N, new_N)
    
    for i in 1:new_N
        for j in 1:new_N
            i_start = (i-1)*factor + 1
            i_end = i*factor
            j_start = (j-1)*factor + 1
            j_end = j*factor
            hic_binned[i, j] = sum(hic[i_start:i_end, j_start:j_end])
        end
    end
    
    return hic_binned
end

function run(output_folder::String; hic_wildcard::String = DEFAULT_HIC_WILDCARD,
             resolution::Int = DEFAULT_RESOLUTION, balanced::Bool = false,
             hic_chrs::Vector{String} = CHR_X_SYNONYMS, cmap::String = DEFAULT_CMAP,
             clim::Vector{Float64} = DEFAULT_CLIM, title::String = DEFAULT_TITLE,
             plot_format::String = DEFAULT_PLOT_FORMAT)
    @info "Plotting HiC for $output_folder with color map: $cmap in file format: $plot_format ..."
    
    if isempty(cmap)
        cmap = DEFAULT_CMAP
    end
    if isempty(plot_format)
        plot_format = DEFAULT_PLOT_FORMAT
    end
    
    pattern = joinpath(output_folder, hic_wildcard)
    hic_files = sort(glob(pattern))
    
    for hic_file in hic_files
        hic = get_hic(hic_file, resolution, balanced, hic_chrs)
        @info "data.shape: $(size(hic))"
        
        hic_log = log10.(replace(hic, 0.0 => 1e-10))
        
        res_kb = resolution / 1000
        n = size(hic_log, 1)
        # Axis formatters: show bin indices as genomic position in kb (like Python FuncFormatter)
        tick_step = max(1, n ÷ 10)
        tick_pos = 1:tick_step:n
        tick_labels = round.(Int, (tick_pos .* res_kb))
        
        # Title: clean hic_file like Python (replace '_hic_003.hdf5', '' and './')
        hic_file_clean = replace(hic_file, "_hic_003.hdf5" => "")
        hic_file_clean = replace(hic_file_clean, "./" => "")
        title_str = replace(title, "{hic_file}" => hic_file_clean)
        title_str = replace(title_str, "{output_folder}" => output_folder)
        title_str = replace(title_str, "{resolution}" => string(resolution))
        title_str = replace(title_str, "{balanced}" => string(balanced))
        
        p = heatmap(hic_log, colormap=get_plots_colormap(cmap), title=title_str,
                    xticks=(tick_pos, string.(tick_labels)),
                    yticks=(tick_pos, string.(tick_labels)))
        
        if length(clim) > 1
            plot!(p, clims=(clim[1], clim[2]))
        end
        
        basename_file = basename(hic_file)
        extension = split(basename_file, ".")[end]
        plot_file = joinpath(output_folder, replace(basename_file, ".$extension" => "_$(cmap).$(plot_format)"))
        
        savefig(p, plot_file; dpi=DEFAULT_DPI)
        @info "Hic plot saved in file $plot_file"
    end
end

function main()
    s = ArgParseSettings()
    @add_arg_table! s begin
        "--output_folder", "-o"
            help = "'Analysis' step output folder containing raw hic_*.hdf5 files"
            default = "."
        "--hic_wildcard", "-w"
            help = "Hi-C files wildcard"
            default = DEFAULT_HIC_WILDCARD
        "--hic_chrs"
            help = "Synonyms of the chromosome from Hi-C matrixes to be plotted"
            nargs = '*'
            default = CHR_X_SYNONYMS
        "--resolution", "-r"
            help = "Hi-C data resolution in bp"
            arg_type = Int
            default = DEFAULT_RESOLUTION
        "--balanced", "-b"
            help = "Get balanced if available, for .cool and .mcool"
            action = :store_true
        "--cmap", "-c"
            help = "Color map"
            default = DEFAULT_CMAP
        "--clim"
            help = "Set the color limits of the current image"
            nargs = '*'
            arg_type = Float64
            default = DEFAULT_CLIM
        "--plot_format", "-f"
            help = "Image file format extension: png, tif, svg"
            default = DEFAULT_PLOT_FORMAT
        "--title"
            help = "Plot title template"
            default = DEFAULT_TITLE
    end
    
    args = parse_args(s)
    
    # Convert vectors from Vector{Any} to proper types
    clim = length(args["clim"]) > 0 ? [Float64(x) for x in args["clim"]] : DEFAULT_CLIM
    hic_chrs = [String(x) for x in args["hic_chrs"]]
    
    run(args["output_folder"],
        hic_wildcard=args["hic_wildcard"],
        resolution=args["resolution"],
        balanced=args["balanced"],
        hic_chrs=hic_chrs,
        cmap=args["cmap"],
        clim=clim,
        title=args["title"],
        plot_format=args["plot_format"])
end

end # module

